### PR TITLE
Allow running of local partitioners on local clusters

### PR DESCRIPTION
**Pull request by thvasilo**
*Issue #, if available:*

*Description of changes:*

Script and implementation that allows us to run distributed partitioning locally and on a local (e.g. EC2) cluster. Original author @classicsong 

### Testing

* Unit test
* Ran compose job using the SageMaker codepath on ML-25M
* Ran local distributed job on ml-100k using a command like:

```bash
python python/graphstorm/gpartition/dist_partition_graph.py --input-path /home/thvasilo/data/ml-100k \
    --metadata-filename updated_row_counts_metadata.json \
    --output-path /tmp/local-part-test/ --num-parts 2 \
    --dgl-tool-path /home/thvasilo/repos/dgl/tools/ \
    --ip-list ip_list.txt --do-dispatch
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docker/build_docker_sagemaker.sh [hunk 1]
```diff
@@ -10,26 +10,12 @@ else
...
-    IMAGE_NAME="$3"
-fi
-
-# process argument 4: image's tag name, default is sm
-if [ -z "$4" ]; then
-    TAG="sm"
-else
-    TAG="$4"
-fi
+# # process argument 2: docker image type, default is GPU
```
Comment (classicsong) [inline docker/build_docker_sagemaker.sh:13]: If you are going to change this script, it is better to create another PR. I don't think you change the functionality of this script.

File: python/graphstorm/gpartition/__init__.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"""
+    Copyright 2023 Contributors
```
Comment (classicsong) [inline python/graphstorm/gpartition/__init__.py:2]:
```suggestion
    Copyright 2024 Contributors
```

File: python/graphstorm/gpartition/dist_partition_graph.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
+"""
+    Copyright 2023 Contributors
```
Comment (classicsong) [inline python/graphstorm/gpartition/dist_partition_graph.py:2]:
```suggestion
    Copyright 2024 Contributors
```

File: python/graphstorm/gpartition/__init__.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+"""
```
Comment (classicsong) [inline python/graphstorm/gpartition/__init__.py:1]: Please add python/graphstorm/gpartition/ into lint test.

File: python/graphstorm/gpartition/random_partition.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+"""
+    Copyright 2023 Contributors
```
Comment (classicsong) [inline python/graphstorm/gpartition/random_partition.py:2]:
```suggestion
    Copyright 2024 Contributors
```

File: python/graphstorm/gpartition/random_partition.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import os
+import logging
+import json
+from typing import Dict, List
+
+import numpy as np
+import pyarrow as pa
+import pyarrow.csv as pa_csv
+
+from .local_partition_graph import LocalPartitionAlgorithm
```
Comment (classicsong) [inline python/graphstorm/gpartition/random_partition.py:27]: where is local_partition_graph?

Comment (thvasilo) [reply python/graphstorm/gpartition/random_partition.py:27]: Sorry file was not added to the commit, I added it now.

File: python/graphstorm/gpartition/local_partition_graph.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+        # Creates the partition metadata file
+        self._create_metadata(num_partitions, partition_assignment_dir)
+        assert os.path.exists(os.path.join(partition_assignment_dir, "partition_meta.json")), \
+            f"Missing partition partition_meta.json file under {partition_assignment_dir}"
+
+    @abstractmethod
+    def _assign_partitions(self, num_partitions: int, partition_dir: str):
+        """Assigns each node in the data to a partition from 0 to `num_partitions-1`,
+        and creates one "{ntype}".json partition assignment file per node type.
```
Comment (classicsong) [inline python/graphstorm/gpartition/local_partition_graph.py:74]: <node_type>.txt or "{ntype}".json?

File: python/graphstorm/gpartition/local_partition_graph.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
+"""
```
Comment (classicsong) [inline python/graphstorm/gpartition/local_partition_graph.py:1]: Shall we rename it to partition_algo.py?
The name local_partition_graph.py is a little bit confusing as we also have dist_partition_graph.py but with totally different impl.

File: python/graphstorm/gpartition/local_partition_graph.py (hunk 7 above)
Comment (thvasilo) [reply python/graphstorm/gpartition/local_partition_graph.py:74]: Good catch thanks!

File: python/graphstorm/gpartition/local_partition_graph.py (hunk 8 above)
Comment (thvasilo) [reply python/graphstorm/gpartition/local_partition_graph.py:1]: Good point, I'll rename to `partition_algo_base.py`

**Review by classicsong — APPROVED** [review]
LGTM

**Pull request merged**